File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        size_t hayLen = haystack.size();
        size_t needleLen = needle.size();

        if (needleLen > hayLen) {
            return -1;
        }

        // Create prefix array
        vector<unsigned int> lps(needleLen);

        lps[0] = 0;
        size_t i = 1;
        size_t length = 0;

        while (i < needleLen) {
            if (needle[i] == needle[length]) {
                ++length;
                lps[i] = length;
                ++i;
            }

            else if (length > 0) {
                length = lps[length - 1];
            }

            else {
                lps[i] = 0;
                ++i;
            }
        }

        // i goes through haystack
        // j goes through needle
        i = 0;
        int j = 0;

        while (i < hayLen) {
            if (haystack[i] == needle[j]) {
                ++i;
                ++j;

                if (j == needleLen) {
                    return i - j;
                }
            }

            else if (j != 0) {
                j = lps[j - 1];
            }

            else {
                ++i;
            }
        }

        return -1;
    }
};

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        hayLen = len(haystack)
        needleLen = len(needle)

        if (hayLen < needleLen):
            return -1
        
        # Set up the LPS array
        lps_array = [0] * needleLen
        lps_array[0] = 0
        index = 1
        length = 0

        while index < needleLen:
            if needle[index] == needle[length]:
                length += 1
                lps_array[index] = length
                index += 1
            
            elif length > 0:
                length = lps_array[length - 1]
            
            else:
                lps_array[index] = 0
                index += 1
        
        # i is for the haystack, j is for needle
        i = 0
        j = 0

        while i < hayLen:
            if haystack[i] == needle[j]:
                i += 1
                j += 1

                if j == needleLen:
                    return i - j
            
            elif j > 0:
                j = lps_array[j - 1]
            
            else:
                i += 1
        
        return -1
